lib/bloc_record/selection.rb

```diff
@@ -1,17 +1,250 @@
+					puts "Invalid id data type"
+				end
+			end
+
+			if ids_in_range && ids_valid
+				rows = connection.execute <<-SQL
+					SELECT #{columns.join ","} FROM #{table}
+					WHERE id IN (#{ids.join(",")});
+				SQL
+
+				rows_to_array(rows)
+			else
+				puts "ids out of range"
+			end
+		end
+	end
+
+
+	def find_one(id)
+		# straight-forward SELECT that returns the row matching the id passed in
+		# By using columns instead of (*) we can manipulate this method later
 		row = connection.get_first_row <<-SQL
 			SELECT #{columns.join ","} FROM #{table}
 			WHERE id = #{id};
 		SQL
 
-		# then we take the returned SQL data, hash it, and return it
+		init_object_from_row(row)
+	end
+
+
+	def find_by(attribute, value)
+		# here we return the first match where attribute == value for the record
+		# we rely on init_object_from_row to return an object
+
+		# first, check for proper data types / not null
+
+		if attribute && attribute.is_a?(Symbol) && value
+		
+			row = connection.get_first_row <<-SQL
+				SELECT #{columns.join ","} FROM #{table}
+				WHERE #{attribute} = #{BlocRecord::Utility.sql_strings(value)};
+			SQL
+
+			init_object_from_row(row)
+		else
+			puts "attribute must be a string and value cannot be null for find_by"
+		end
+	end
+
+
+	def take_one
+		# Here we are just grabbing a random row from the table
+		row = connection.get_first_row <<- SQL
+			SELECT #{columns.join ","} FROM #{table}
+			ORDER BY random()
+			LIMIT 1;
+		SQL
+
+		init_object_from_row(row)
+	end
+
+
+	def take(num=1)
+		# similar to take_one() but with a variable number of return random items as an array
+		# if single num or no num passed in, returns object courtesy take_one()
+
+		# let's make sure num is an integer
+		# if it is an integer, lets make sure it is a positive number and less than set count
+
+		if num.is_a? Integer
+			if num > 0 && num <= self.count
+				if num > 1
+					rows = connection.execute <<-SQL
+						SELECT #{columns.join ","} FROM #{table}
+						ORDER BY random()
+						LIMIT #{num};
+					SQL
+				else
+					take_one
+				end
+			else
+				puts "Number of items to take is out of range"
+			end
+		else
+			puts "Argument is not a valid number"
+		end
+	end
+
+	# first() and last() return first or last object in the database
+	def first
+		row = connection.get_first_row <<-SQL
+			SELECT #{columns.join ","} FROM #{table}
+			ORDER BY id
+			ASC LIMIT 1;
+		SQL
+
+		init_object_from_row(row)
+	end
+
+	def last
+		row = connection.get_first_row <<-SQL
+			SELECT #{columns.join ","} FROM #{table}
+			ORDER BY id
+			DESC LIMIT 1;
+		SQL
+
+		init_object_from_row(row)
+	end
+
+
+	def all
+		rows = connection.execute <<-SQL
+			SELECT #{columns.join ","} FROM #{table};
+		SQL
+
+		rows_to_array(rows)
+	end
+
+
+	def method_missing(method, *arguments, &block)
+		if method == "find_by_name"
+			find_by(:name, *arguments)
+		else
+			puts "no such method found"
+		end
+	end
+
+	def find_each(attribute, value, batch_size=nil, &block)
```

Comment:
Look again at the example usages. In particular, look at what they pass in for arguments. One thing you will need to remember is what gets passed into a method when a syntax this this is used: `some_object.some_method(some: "thing", some_other: 42)`. If you are not familiar with that syntax, review it; it is a Ruby shorthand syntax. If you end up working a job with Ruby (with or without Rails), I guarantee you will see it.